### PR TITLE
Bug: add conversion from slice to array for selecting rows in pyarrow `__getitem__` 

### LINKED ISSUE #971: [Bug]: Pyarrow `__getitem__` fails with integer for column position
### Describe the bug

Unlike pandas-like and polars, pyarrow backend fails to return a series when sliced with column position integer

### Steps or code to reproduce the bug

```py
import narwhals as nw
import pandas as pd
import polars as pl
import pyarrow as pa

data = {"x1": [0, 1, 2], "x2": [.1, .2, .3]}

@nw.narwhalify
def func(df):
    return df[:, 1]
```

**polars**:

```py
func(pl.DataFrame(data))
```
```terminal
shape: (3,)
Series: 'x2' [f64]
[
	0.1
	0.2
	0.3
]
```

**pandas**

```py
func(pd.DataFrame(data))
```
```terminal
0    0.1
1    0.2
2    0.3
Name: x2, dtype: float64
```

**pyarrow**

```py
func(pa.table(data))
```
> TypeError: Got unexpected argument type <class 'slice'> for compute function

### Expected results

Return series by index position

### Actual results

Raise an error

### Please run narwhals.show_version() and enter the output below.

```shell
System:
    python: 3.12.2 (main, Feb 25 2024, 16:35:05) [GCC 11.4.0]
    machine: Linux-5.15.133.1-microsoft-standard-WSL2-x86_64-with-glibc2.35

Python dependencies:
     narwhals: 1.8.0
       pandas: 2.2.2
       polars: 1.7.1
         cudf: 
        modin: 
      pyarrow: 17.0.0
        numpy: 2.1.1
```

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by raisadz**
## What type of PR is this? (check all applicable)

- [ ] 💾 Refactor
- [ ] ✨ Feature
- [x] 🐛 Bug Fix
- [ ] 🔧 Optimization
- [ ] 📝 Documentation
- [ ] ✅ Test
- [ ] 🐳 Other

## Related issues 

- Related issue # 
- Closes #971

## Checklist

- [x] Code follows style guide (ruff)
- [x] Tests added 
- [ ] Documented the changes

### CODE REVIEW

**Review by MarcoGorelli — APPROVED** [review]
awesome! thanks @raisadz !

File: narwhals/_arrow/dataframe.py [hunk 1]
```diff
@@ -174,13 +179,22 @@ def __getitem__(
                     )
                 msg = f"Expected slice of integers or strings, got: {type(item[1])}"  # pragma: no cover
                 raise TypeError(msg)  # pragma: no cover
-
             from narwhals._arrow.series import ArrowSeries
 
             # PyArrow columns are always strings
             col_name = item[1] if isinstance(item[1], str) else self.columns[item[1]]
+            assert not isinstance(item[0], str)  # help mypy  # noqa: S101
```
Comment (FBruzzesi) [inline narwhals/_arrow/dataframe.py:186]: Fancy changing this from `assert` to a `raise` statement as a follow up?

[the dangers of assert in python](https://snyk.io/blog/the-dangers-of-assert-in-python/)

TL;DR: `assert` is only run in debug mode, if you run python with the optimized flag, that will be skipped.

Comment (raisadz) [reply narwhals/_arrow/dataframe.py:186]: Thanks for your comment @FBruzzesi , I replaced `assert` with `raise` in this PR https://github.com/narwhals-dev/narwhals/pull/980

Please, note that the line never gets hit so `# pragma: no cover` is added

Comment (FBruzzesi) [reply narwhals/_arrow/dataframe.py:186]: Oh I see, thanks for pointing that out 👌! Let's move the conversation there anyway
